packages/@aws-cdk/cfn-include/lib/cdk-include.ts

```diff
@@ -0,0 +1,221 @@
+  private readonly cfnType2L1Class: { [cfnType: string]: string };
+  private readonly template: any;
+  private readonly preserveLogicalIds: boolean;
+
+  constructor(scope: core.Construct, id: string, props: CdkIncludeProps) {
+    super(scope, id);
+
+    // read the template into a JS object
+    this.template = fs.readJsonSync(props.templateFile);
+
+    // check for any sections of the template we don't recognize
+    for (const section of Object.keys(this.template)) {
+      switch (section) {
+        case 'AWSTemplateFormatVersion':
+        case 'Description':
+        case 'Resources':
+          // currently, we only support these 3 sections
+          break;
+        default:
+          throw new Error(`Encountered an unsupported section in the template: '${section}'`);
+      }
+    }
+
+    // read the map of CloudFormation type to their L1 class created by the build script
+    this.cfnType2L1Class = fs.readJsonSync(path.join(__dirname, '..', 'cfn-types-2-classes.json'));
+    this.preserveLogicalIds = props.preserveLogicalIds ?? true;
+
+    // instantiate all resources as CDK L1 objects
+    for (const logicalId of Object.keys(this.template.Resources || {})) {
+      this.getOrCreateResource(logicalId);
+    }
+  }
+
+  /**
+   * Returns the low-level CfnResource from the template with the given logical ID.
+   * Any modifications performed on that resource will be reflected in the resulting CDK template.
+   *
+   * The returned object will be of the proper underlying class;
+   * you can always cast it to the correct type in your code:
+   *
+   *     // assume the template contains an AWS::S3::Bucket with logical ID 'Bucket'
+   *     const cfnBucket = cfnTemplate.getResource('Bucket') as s3.CfnBucket;
+   *     // cfnBucket is of type s3.CfnBucket
+   *
+   * If the template does not contain a resource with the given logical ID,
+   * an exception will be thrown.
+   *
+   * @param logicalId the logical ID of the resource in the CloudFormation template file
+   */
+  public getResource(logicalId: string): core.CfnResource {
+    const ret = this.resources[logicalId];
+    if (!ret) {
+      throw new Error(`Resource with logical ID '${logicalId}' was not found in the template`);
+    }
+    return ret;
+  }
+
+  private getOrCreateResource(logicalId: string): core.CfnResource {
+    const ret = this.resources[logicalId];
+    if (ret) {
+      return ret;
+    }
+
+    const resourceConfig: any = this.template.Resources[logicalId];
+    const l1ClassFqn = this.cfnType2L1Class[(resourceConfig as any).Type];
+    if (l1ClassFqn) {
+      const [moduleName, ...className] = l1ClassFqn.split('.');
+      const module = require(moduleName);
+      const jsClassFromModule = module[className.join('.')];
+      // translate the properties from the template to the casing expected by CDK objects
+      const l1Props = jsClassFromModule.cfnProps2CdkProps((resourceConfig as any).Properties);
```

Comment:
Looks like we won't be able to support CFN templates that use properties that aren't in the spec yet, huh?

This is unfortunately not hypothetical, this shit happens all the time when new services generate a lot of interest but the service team is lagging behind in updating the spec (see WAFv2).

I understand this is a !@#$%TY hard problem, but I feel it's important to have a bypass to roundtrip the CloudFormation template as-is. Otherwise, this module is not really a replacement for `CfnInclude`.

Now, we have to decide whether that's okay or not. Instead of `cfn-include` it could be called `cfn-import` or `cfn-parse` and it's clear that it will operate on the structured data, and may fail.

If it's ultimately the vision for this package to pass-through arbitrary templates AND allow structured access, that's fine, and it's also fine if you want to postpone that for now (although I would suggest the order to work in would then be "generic pass-through first, smart parsing second", because otherwise you're setting yourself up for getting to an 80% solution and then having to rewrite everything to achieve the final 20%).

If it's *not* the vision for this package to pass-through arbitrary templates, but to make it more convenient to work with resources defined in a subset of templates that we deign to support, then that's fine too. There is still value in this! 

But I feel we have to clear and crisp on the definition, and choose naming appropriately to set correct user expectations here, and add wording to the README to the same effect.

Reply:
I'm fine with the initial version not supporting perfect pass through of all properties (including ones not in the schema). It's already on my radar to add support for these in the future.